core/src/main/java/org/apache/accumulo/core/clientImpl/KeyExtentCache.java

```diff
@@ -0,0 +1,131 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.clientImpl;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.PartialKey;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.TableId;
+import org.apache.accumulo.core.dataImpl.KeyExtent;
+import org.apache.accumulo.core.metadata.schema.TabletMetadata;
+import org.apache.accumulo.core.metadata.schema.TabletsMetadata;
+import org.apache.accumulo.core.util.Pair;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.benmanes.caffeine.cache.CacheLoader;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.github.benmanes.caffeine.cache.Scheduler;
+
+public class KeyExtentCache {
+
+  private static class KeyExtentCacheLoader
+      implements CacheLoader<Pair<TableId,Range>,List<KeyExtent>> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KeyExtentCacheLoader.class);
+
+    private final ClientContext ctx;
+
+    public KeyExtentCacheLoader(ClientContext ctx) {
+      this.ctx = ctx;
+    }
+
+    @Override
+    public @Nullable List<KeyExtent> load(Pair<TableId,Range> pair) throws Exception {
+      final TableId tid = pair.getFirst();
+      final Range r = pair.getSecond();
+      LOG.trace("Loading extents for tid: {} and range: {}", tid, r);
+      TabletsMetadata tm = ctx.getAmple().readTablets().forTable(tid)
+          .overlapping(r.getStartKey() == null ? null : r.getStartKey().getRow(),
+              r.isStartKeyInclusive(), r.getEndKey() == null ? null : r.getEndKey().getRow())
+          .build();
+      List<KeyExtent> result = new ArrayList<>();
+      for (TabletMetadata t : tm) {
+        if (t.getExtent().endRow() == null || t.getExtent().endRow() != null
+            && !r.afterEndKey(new Key(t.getExtent().endRow()).followingKey(PartialKey.ROW))) {
+          result.add(t.getExtent());
+        }
+      }
+      LOG.trace("Returning extents: {}", result);
+      return result;
+    }
+
+    @Override
+    public Map<? extends Pair<TableId,Range>,? extends List<KeyExtent>>
+        loadAll(Set<? extends Pair<TableId,Range>> pairs) throws Exception {
+
+      // Sort the inputs
+      TreeMap<TableId,SortedSet<Range>> inputs = new TreeMap<>();
```

Comment:
You might be able to use a TreeMultimap here. Not sure if it would make much of a difference though.